feat: tests for cases when data are in both RUB and OS

NGA-TRAN opened the pull request:
Closes #1333 

These are tests and related bug fixes for cases when data is in both RUB and OS. This PR will serve both #1333 and a part of #1082.

# Changes:
  . Added more scenarios in which data is in both RUB and OS that are automatically included in almost all tests under server/query_tests
  . Fixed a bug that always snapshot a RUB chunk if they are both in RUB and OS
  . Fixed a few other related bugs that return correct results if we read data in OS. Note that since we always read from RUB in this PR,  we do not expose read data from parquet files which will be done next.


- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

NGA-TRAN commented on server/src/db/chunk.rs:
```diff
@@ -107,10 +107,37 @@ impl DbChunk {
+                chunk: Arc::clone(chunk),
+                partition_key,
+                // Since data exists in both read buffer and object store, we should
+                // snapshot the chunk of read buffer
```
This is the fix to always read RUB if data exists in both RUB and OS

NGA-TRAN commented:
There are a lot of lines od code commented out. I might need them or will delete them when I finish ~1082

alamb commented:
I plan to review this carefully tomorrow morning

NGA-TRAN commented:
Also, currently we only compare the expected and actual output the see if we return the right results in the tests. I know each of us do debug and verify that the data is read from the right MUB or RUB or OS but there is no auto way in the tests to verify that. We might want to add that in. Maybe debug log it and assert_eq! the log message. What do you think @alamb , @e-dard, and @tustvold ? This will make ourdaily life easier and no longer concerned wrong buffer will be read (e.g. suboptimal OS instead of the optimal RUB) in the future due to some bugs

alamb commented on server/src/query_tests/scenarios.rs:
```diff
@@ -61,28 +68,78 @@ impl DbSetup for NoData {
 
-        let scenario3 = DbScenario {
+        let scenario4 = DbScenario {
             scenario_name: "Empty Database after drop chunk".into(),
```
This name probably should be updated to something like "empty database after drop chunk from both read buffer and object store")

alamb commented on server/src/query_tests/scenarios.rs:
```diff
@@ -356,7 +413,28 @@ pub(crate) async fn make_one_chunk_scenarios(partition_key: &str, data: &str) ->
+        db,
+    };
+
+    // TODO: Add scenario 5 where data is in object store only
```
👍

alamb reviewed (commented):
The tests look good @NGA-TRAN  . I am curious what was wrong when using the async version of the parquet reading code? It would be good to understand that before commenting it all out I think

NGA-TRAN commented on server/src/query_tests/scenarios.rs:
```diff
@@ -61,28 +68,78 @@ impl DbSetup for NoData {
+        //
+        let db = make_db().db;
+        let data = "cpu,region=west user=23.2 100";
+        write_lp(&db, data);
```
And I do add more scenarios for data case else where

crepererum replied:
Ah, that makes sense. Thanks! :)

NGA-TRAN commented on parquet_file/src/storage.rs:
```diff
@@ -292,50 +276,158 @@ impl Storage {
-            }
-        });
+        // TODO: These commented-out code lines will either be used or deleted when #1082 done
+        // TODO: Until this read_filter is an async, we cannot make this multi-threaded yet
```
@alamb  I think here is problem when I tried to debug it. Since the parent's function, read_filter, is not an async, we cannot call await to wait for this to finish. Hence, while this task is still running in the middle and not return results yet, its caller continues and assert_eq!(expected, results) but the results are still empty and causes the tests failed.

At first I was so sure it was the problem. I tried to convert the read_filter to async but since it is a function of a trait applied to all MUB, RUB and OS, I do not want to change it yet and do this. However, after I modified this to not spawning a new thread, I still got tests failed but since a lot of them, I have not had time to investigate whether it is a new problem or this change does not actually help.

If you think my explanation sounds reasonable, I will keep this commented out and investigate it when I have unload_chunk function available. It is hard to put time on this right now because all tests read RUB instead of OS.

alamb replied:
@NGA-TRAN  I think we should just keep moving forward; If you would like I can take a look into figuring out what was going on with no results coming from async.

alamb replied:
I thought more about this last night -- I think getting parquet to a working state is the goal now, and given we didn't have any test coverage of the commented out code anyways we don't know if it worked at all. Thus, commenting it out / removing it for now and building it back up along with tests seems like a good idea to me

NGA-TRAN replied:
Thanks @alamb. Yes, that is what I thought. My plan is to work on Unload RUB to have ObjectStoreOnly state then I will turn this on and look into it carefully. I will ask for help if I do not figure out the problem.

NGA-TRAN commented on parquet_file/src/storage.rs:
```diff
@@ -292,50 +276,158 @@ impl Storage {
+            &mut batches,
+            limit,
+        ) {
+            println!("Parquet reader terminated due to error: {:?}", e);
```
Will do

NGA-TRAN commented:
>   I am curious what was wrong when using the async version of the parquet reading code? It would be good to understand that before commenting it all out I think

 @alamb I have addressed the comments and also replied to the related comment explaining my understanding. Let me know what you think. If you do not think the async is the solution, I will convert it to code back. The main reason I keep them as comments because I do want to use them but I need more tests to prove they do work.
